src/Surfnet/StepupGateway/ApiBundle/Service/U2fVerificationService.php

```diff
@@ -0,0 +1,256 @@
+
+            throw new RuntimeException($errorMessage, 0, $e);
+        }
+
+        $apiResult = new U2fAuthenticationVerificationResult();
+
+        if ($result->wasSuccessful()) {
+            $this->logger->notice('U2F device authentication verification successful');
+            $apiResult->status = U2fAuthenticationVerificationResult::STATUS_SUCCESS;
+        } elseif ($result->didDeviceReportAnyError()) {
+            $apiResult->status = U2fAuthenticationVerificationResult::STATUS_DEVICE_ERROR;
+            $this->logger->error('U2F device reported an error');
+        } elseif ($result->didResponseChallengeNotMatchRequestChallenge()) {
+            $apiResult->status = U2fAuthenticationVerificationResult::STATUS_REQUEST_RESPONSE_MISMATCH;
+            $this->logger->error('Response challenge did not match request challenge');
+        } elseif ($result->wasRegistrationUnknown()) {
+            $apiResult->status = U2fAuthenticationVerificationResult::STATUS_REGISTRATION_UNKNOWN;
+            $this->logger->error('Registration unknown');
+        } elseif ($result->wasResponseNotSignedByDevice()) {
+            $apiResult->status = U2fAuthenticationVerificationResult::STATUS_RESPONSE_NOT_SIGNED_BY_DEVICE;
+            $this->logger->error('Response was not signed by device');
+        } elseif ($result->didPublicKeyDecodingFail()) {
+            $apiResult->status = U2fAuthenticationVerificationResult::STATUS_PUBLIC_KEY_DECODING_FAILED;
+            $this->logger->error('Decoding of the public key failed');
+        } elseif ($result->didntAppIdsMatch()) {
+            $apiResult->status = U2fAuthenticationVerificationResult::STATUS_APP_ID_MISMATCH;
+            $this->logger->critical('The AppID of a U2F message didn\'t match the server\'s');
+        } else {
+            throw new LogicException('Unknown authentication verification result status');
+        }
+
+        return $apiResult;
+    }
+
+    /**
+     * @param RevokeRequest $revokeRequest
+     * @param Requester $requester
+     */
+    public function revokeRegistration(RevokeRequest $revokeRequest, Requester $requester)
+    {
+        $this->logger->notice('Received request to revoke a U2F device registration');
+
+        try {
+            $this->verificationService->revokeRegistration(new KeyHandle($revokeRequest->keyHandle));
+        } catch (Exception $e) {
+            $errorMessage = sprintf(
+                'An exception was thrown while revoking the U2F device registration (%s: %s)',
+                get_class($e),
+                $e->getMessage()
+            );
+            $this->logger->critical($errorMessage, ['exception' => $e]);
+
+            throw new RuntimeException($errorMessage, 0, $e);
+        }
+
+        $this->logger->notice('Revoked U2F device registration');
+    }
+
+    /**
+     * @param U2fRegisterRequest $request
+     * @return RegisterRequest
+     */
+    private function adaptRegisterRequest(U2fRegisterRequest $request)
```

Comment:
seeing all the conversions everywhere, might consider making an adapter object which then can also be tested for correctness.